### ISSUE TITLE
Convert components to use EnabledComponent mixin

### ISSUE THREAD

**Issue by zepumph** [phetsims/sun#585]
From https://github.com/phetsims/sun/issues/257, once that issue is complete, and EnabledComponent is "production ready", we should begin swapping out custom enabledProperty instances for this mixin. Hopefully each of these will be minimally invasive, because the api should likely not change (in most cases at least). 

Here is a list of components to update. Most probably deserve dedicated issues, but this issue can serve as a central location for the status of the conversion as a whole.

sun:
- [x] AquaRadioButtonGroup.js (addition not conversion, see https://github.com/phetsims/sun/issues/582)
- [x] AquaRadioButton.js
- [x] Checkbox.js
- [x] ComboBox.js
- [x] NumberSpinner.js
- [x] Slider.js
- [x] RadioButtonGroup.js
- [x] ButtonModel.js (this one may be a bit tricky, waiting on the outcome of the creation issue to see if this is supported)
- [x] RoundButtonView.js and RectangularButtonView. These likely will use the solution for ButtonModel, but maybe they will want their own?

scenery-phet:
- [x] EyeDropperNode
- [x] FineCoarseSpinner
- [x] GaugeNode
- [x] LaserPointerNode
- [x] NumberControl
- [x] NumberPicker
- [x] TimeControlNode

tambo:
- [x] AmplitudeModulator

I will mark this for dev meeting once this is ready.

**Comment by zepumph** [comment]
One thing that we need to make sure to do as part of regression testing is to look at interrupting input when toggling enabled. This has been added to EnabledComponent, but isn't is most custom usages across the project. 

over in https://github.com/phetsims/sun/issues/257#issuecomment-599724039 @jbphet  said:
> There are TODO items about whether to call interruptSubtree and whether to set the cursor in the handler for changes to the enabled state. Both seem like good ideas to me. I'd suggest doing them, and setting up a QA task to test this - with an emphasis on multi-touch - to see whether there are any subtle implications that aren't immediately obvious to us.

I agree, but think it will be best to do this as a batch here (or in component specific sub issues) when components are converted as opposed to in https://github.com/phetsims/sun/issues/257.

**Comment by zepumph** [comment]
From https://github.com/phetsims/sun/issues/257#issuecomment-614310370:

- [ ] Note that when converting `AquaRadioButton.js`, the api for it has changed, as we don't support a default value passed in via `options.enabled`. If that proves too challenging, we can add it, but it is nice to try to keep a simpler api to begin with.

**Comment by pixelzoom** [comment]
@ariel-phet asked me to take this on. The goal is to have `enabledProperty` supported uniformly across PhET UI components, so that `opacityProperty` and `pickableProperty` can be (by default) uninstrumented for PhET-iO.  

I will be looking through other repos for similar/related issues, and will self-assign as I find them.

**Comment by zepumph** [comment]
For clarity, those phet-io Properties are scheduled for uninstrumentation (by default) over in https://github.com/phetsims/scenery/issues/1047, and I do not think of this issue's work as a precursor to that. Most cases already have an enabledProperty specified that is PhET-iO instrumented, and so the conversion over here will not actually change the API significantly (or at all). 

For PhET-iO, I am not aware of cases where we have elected to tell the client to use opacity/pickable instead of manually adding an instrumented `enabledProperty` to come UI component. 

This comment shouldn't change your workflow, I say this only to increase understanding of this work being done and the dependencies (or lack there of) that it has on each other.

**Comment by pixelzoom** [comment]
Yikes. While working on another issue, I discovered that in addition to `EnabledComponent`, there is `EnabledNode` and `EnabledPhetioObject`.   They are all documented the same, all authored by @zepumph, and all updated recently.   @zepumph can you please clarify (or better yet document) the relationship between these 3 classes, and when each one should be used?

**Comment by zepumph** [comment]
This is out for review over in #257 and I responded to this there, see https://github.com/phetsims/sun/issues/257#issuecomment-700358084. This issue should be on hold until that issue has been reviewed. I did work here for a specific PhET-iO requirement. Marking back on hold until that issue is reviewed.

**Comment by pixelzoom** [comment]
Conversion is done. Summary of the above commits:

* EyeDropperNode was using `enabledProperty` incorrectly, to change the state of its button.  So I renamed it to `buttonEnabledProperty`.

* LaserPointerNode was using `enabledProperty` incorrectly, to change the state of its button. This feature was unused except in scenery-phet demo, so I simply deleted it.

* GaugeNode was using `enabledProperty` incorrectly, to control whether the gauge updated when invisible. I replaced it with `options.updateWhenInvisible`.

* All of the other other classes in the  https://github.com/phetsims/sun/issues/585#issue-588579232 checklist are now using EnabledComponent (for model components) or EnabledNode (for view components, Node subclasses).

@ariel-phet would you like to have someone review? Perhaps @zepumph since he wrote the EnabledComponent stack?

**Comment by zepumph** [comment]
I'll take a look since I got to this issue anyways while working on https://github.com/phetsims/scenery/issues/1047.

**Comment by zepumph** [comment]
I have no problem with https://github.com/phetsims/sun/commit/c3f449b7542ad484b7a87ee48600360372c67a31, but was wondering if you could explain your rationale.

**Comment by zepumph** [comment]
I looked through every commit, I also checked out the new features in the scenery-phet demo. I then ran a couple of sims that used  these components and played around with setting their enabledProperties, and the same when passing in an `enabledProperty`. 

In general I am totally ecstatic about these changes. The above  commits removed so much duplication! Thank you so much @pixelzoom for taking this on. This may be the final manifestation of how we support enabled on components at PhET, or things may develop further in https://github.com/phetsims/scenery/issues/1100. Either way this issue was a huge step towards our end goal. Thanks! 


Back to you.

**Comment by zepumph** [comment]
>Test interruptSubtree regression with QA

@pixelzoom can you also please comment on this, do you think this is necessary? If so how should we set that up? What tests may be good to do?

**Comment by pixelzoom** [comment]
> I have no problem with c3f449b and others like it, but was wondering if you could explain your rationale.

AccessibleSlider uses the enabledProperty created by EnabledNode.  So `initializeEnabledNode` must be called before `initializeAccessibleSlider`.  Therefore it seems like good hygiene to call `mixInto` in the same order.

**Comment by pixelzoom** [comment]
> >- [ ] Test interruptSubtree regression with QA
> 
> @pixelzoom can you also please comment on this, do you think this is necessary? If so how should we set that up? What tests may be good to do?

In general, developers and QA should be verifying that "in progress" interactions that are cancelled when the target of the interaction becomes disabled.  I doubt that's happening, and I don't know how to make that happen other than a PSA/reminder. 

If there are any regressions introduced, they might be caught by CT, assuming that fuzzing uses multitouch.

**Comment by zepumph** [comment]
>If there are any regressions introduced, they might be caught by CT, assuming that fuzzing uses multitouch.

That will be added as part of https://github.com/phetsims/aqua/issues/106.


>In general, developers and QA should be verifying that "in progress" interactions are cancelled when the target of the interaction becomes disabled. I doubt that's happening, and I don't know how to make that happen other than a PSA/reminder.

I agree, but also know that multi-touch testing is often not done (for cost saving). Perhaps a discussion point on the dev meeting agenda? I would also be fine with catching things in CT and calling it good. Up to you!

**Comment by pixelzoom** [comment]
> I agree, but also know that multi-touch testing is often not done (for cost saving). 

Not done by whom? Developers, QA, both?  Imo there's not much cost if the developer checks this "as you go" -- it's just a good habit.  And it shouldn't be QA's responsibility to do all of the multi-touch testing.

**Comment by pixelzoom** [comment]
Assigning to @ariel-phet to decide if he wants to do something about (a) testing for `interruptSubtreeInput` regressions, and (b) clarifying (in general) what multi-touch testing should be done and by whom.

**Comment by zepumph** [comment]
> > I agree, but also know that multi-touch testing is often not done (for cost saving).
> 
> Not done by whom? Developers, QA, both? Imo there's not much cost if the developer checks this "as you go" -- it's just a good habit. And it shouldn't be QA's responsibility to do all of the multi-touch testing.

In that message I was speaking to QA personally. I know that a lot of devs don't have access to multi-touch devices, and it likely isn't in many people's workflow (it definitely isn't in mine). Perhaps a psa at dev meeting would be nice.

**Comment by ariel-phet** [comment]
@pixelzoom I think doing a regression test would be good. Can you make a QA issue with some suggestions of what to test/how (not immediately clearly to me) and place it in the pipeline project (I will triage, but I think there is room to test immediately)

**Comment by pixelzoom** [comment]
Discussed on Zoom with @ariel-phet.  I really don't know how to specify this as a task for QA.  So what we decided is:  I'll identify which classes need to be tested, create an issue in QA repo, and assign to @ariel-phet.  That issue is https://github.com/phetsims/QA/issues/572.

Closing.
